### DISCUSSION TITLE
Design of the Profile Display

### DISCUSSION THREAD

**Discussion by CodedOre** [CodedOre/NewCaw#21] [Design]
As I have merged in the `Profile` class (containing extended data for a user) today, I think it's time to discuss how they should be displayed in the user detail page. In fact, I think we did not discussed the design for the user overview yet...

I have created a prototype today, but would like to get some feedback before implementing it.

<img src="https://user-images.githubusercontent.com/47981497/144276504-bd18d7d7-3492-4af0-a013-4126c9a7ae24.png" width=32% height=32%> <img src="https://user-images.githubusercontent.com/47981497/144276558-930b899f-5b7c-44d3-b7cc-e82034167455.png" width=60% height=60%>

**Comment by IBBoard** [comment]
Looks good to me. All the bits I'd expect on the profile data - avatar, background, name, handle, bio, joined date, location, website.

I thought it might be nice to put the follow button in the white space on the wider one, but a) that involves it jumping around in weird ways between wide and narrow views and b) I'm sure there will be places where it is in wide mode but the user's display name is long enough that there's not much white space to use!

I think we'd need a different icon for "joined" though. That's the RSS feed icon. Or Windows wifi, but flipped horizontally. Not sure what we'd use, though. Calendar?

I was going to comment on the lack of "Favourited" list, but it looks like we don't have that in Cawbird anyway 🙂 I think we can do Twitter's "Tweets" and "Tweets and Replies" with a toggle in the Posts view. I wouldn't bother with Twitter's "Media" (partly because there's not a single API for it and I think we'd have to rely on Search with its 7 day history, and partly because it's a low priority feature). Followers, Following and Lists are probably still the most useful.

**Comment by CodedOre** [comment]
Alright, got some feedback from the Gnome Design team.

One of the first points was the switcher for posts, follower, etc.
While Adwaita might get a dedicated design for [inline view switcher](https://gitlab.gnome.org/Teams/Design/os-mockups/-/issues/141) which we could use, it was brought up that the categories Followers, Follows and Lists "aren't equal nor homogeneous", so it could make sense to place the access to those elsewhere.

Another point was to make the metadata more subtle and drop the list.
While I personally don't think it looks bad, it does consume a lot of space, so it probably makes sense to replace it with something more lightweight.

With both of these done, the prototype would look like this:

<img src="https://user-images.githubusercontent.com/47981497/144444672-1a3911a5-953a-4b60-931b-e2c20c016653.png" width=32% height=32%> <img src="https://user-images.githubusercontent.com/47981497/144444663-0d57da4c-e69c-467e-9b68-26637419f7ac.png" width=60% height=60%>

**Reply by CodedOre** [reply]
Got even more different:

![ProfileDisplay without headerbar](https://matrix-client.matrix.org/_matrix/media/r0/download/matrix.org/oYTqPvEVIkwKXTaLYoccrTwJ)

**Reply by IBBoard** [reply]
They've got a point (which is why they're a design team!).

I think a mix of those two might be best - I prefer the name next to the avatar on wider displays because there's going to be a lot of empty space in that second design. But I can see why they don't want to put a button over the cover picture.

The details including follower count make more sense below the bio rather than above the bio, because it keeps "stuff the user wrote" (name and bio) and "bulleted stuff about the user" (followers, joined date, location, URL) together in their own groups.

Would timeline and lists make more sense as different views now? Lists aren't entirely equivalent to a timeline, but the list does contain a timeline, so it's only one step away and I can't see a better place to put it. The `⋮` button wouldn't be a place that I'd look for lists, because it's for profile-related actions (block, mute, etc). A third view could then be likes (although I suspect the Twitter v2 API might burn a lot of tweet credits on that!)

Also, I think split buttons are out of favour in design at the moment and separate "more" buttons are better where possible. It probably makes sense for accessibility.

**Comment by jannuary** [comment]
Some personal design exploration:
![Slice 2](https://user-images.githubusercontent.com/27908024/144745906-e55d5a89-e9a5-4a40-97ac-a69b477d1bcf.png)

**Reply by IBBoard** [reply]
Interesting. I'm not keen on "header bar disappears" - it's fine for mobile but desktop doesn't have the same edge of the device and so it'll look odd and out of place. The "pills" could be a good way to do the filtering of "don't show RTs" etc, but "Likes" is a separate feed and needs a separate list. We could theoretically fetch both feeds and merge the lists BUT a) it'll make the UI unclear, b) likes *are not* RTs and shouldn't be treated like them! (that's part of the reason we're using Cawbird and not Twitter Web!) and c) it's an extra request that counts towards the tweet cap in the v2 API, so we should avoid making that call if people don't want to see it.

**Reply by jannuary** [reply]
I mean, it's not a matter of the header bar "disappearing" - the header image is basically kind of an oversized headerbar in of itself, and would just collapse to a regular headerbar when scrolling.

About likes, yeah I suppose that's true, but otherwise I'm not sure how would showing likes be exposed in UI, if at all.

**Reply by IBBoard** [reply]
I guess "disappearing" is the wrong word. There's no system-themed header bar, which this kind of app still needs on desktop.

I think the likes are going to have to be a separate list in a stack - one for the timeline tweets, one for liked tweets, and one for lists (which then becomes a list of tweets when you pick a list). It's not 100% "equal and homogenous" but it's close.

**Comment by CodedOre** [comment]
I now have (mostly) implemented the UI for the profile information:

![current-profile-display](https://user-images.githubusercontent.com/47981497/145839336-98b8a0eb-6efe-4007-997a-26c19be80936.png)

One thing is missing currently, which is the profile header, which opens a question:
Until now, I have just assumed that we keep the image placeholder at a static height of 200px. While this makes the rest of the UI easier to manage, it would result in the picture being cropped to some degree, as the aspect ration changes the more width an window has, while the one of the image (15:5) stays the same.

Question now is: Do we accept the cropping and allow to click on the profile to open an MediaDisplay with the profile header? Or do we want to keep the header completely visible no matter what?

**Reply by IBBoard** [reply]
Twitter seems to keep it proportional (although you've got to get REALLY narrow before it actually changes the column width).

I think we've got to crop it. I think people are used to it. Clicking on it might be a nice to have (same as the avatar - click to get the bigger one). And if it gets too wide, do we have a max width for the content or do we need to have some kind of blurring effect (like news channels do when showing someone's portrait oriented phone footage on a widescreen TV)?

Does that work for Mastodon as well, or does it have different sizes?

Also, can we have a proper header bar on the profile, please? We want the text to be readable whatever ugly-arse image people use for a cover, and the easiest way is to not overlap the two 😉

**Reply by CodedOre** [reply]
> And if it gets too wide, do we have a max width for the content or do we need to have some kind of blurring effect (like news channels do when showing someone's portrait oriented phone footage on a widescreen TV)?

I would argue for blurring, it's just the question how to accomplish that...

> Does that work for Mastodon as well, or does it have different sizes?

Mastodon uses the same aspect ratio there. If I've read the docs correctly, they also scale all headers to 1500x500.

**Reply by CodedOre** [reply]
> Also, can we have a proper header bar on the profile, please? We want the text to be readable whatever ugly-arse image people use for a cover, and the easiest way is to not overlap the two 😉

It's where I landed after some discussion with the design team. 

Going to put the relevant discussion from the Matrix room here ASAP.

**Reply by IBBoard** [reply]
> It's where I landed after some discussion with the design team.

But we're a desktop app first, not a phone app. Desktop apps have desktop widgets. Like a title bar! Suddenly having the header bar disappear is going to look *really* strange and out of place on a desktop app. The only time I can see it making sense is for edge-to-edge gaming (IIRC ones of the emulators does it).

**Reply by CodedOre** [reply]
If you ask me the light header bar (with only back and window control buttons) contrasted with an image directly beneath is not necessary a best design.

Here is the discussion with the design team about this:

![Part 1](https://user-images.githubusercontent.com/47981497/145887261-1fc4cb18-692e-4eda-a94e-cfcce64b6179.jpeg)
![Part 2](https://user-images.githubusercontent.com/47981497/145887569-b358dfa0-c107-4ee0-b66e-10bb9f687021.jpeg)
![Part 3](https://user-images.githubusercontent.com/47981497/145887598-9fe78a5d-1185-4528-8475-ae64f47b23de.jpeg)
![Part 4](https://user-images.githubusercontent.com/47981497/145887625-92e09838-56f2-43ff-88ff-85bb92588eea.jpeg)
![Part 5](https://user-images.githubusercontent.com/47981497/145887644-532c0ce0-0fa6-4a28-91ec-1cf0754ea8c3.jpeg)

Also the links that was posted by Exalm: https://twitter.com/gilbouhnick/status/1094717120909398016?s=21 https://www.b4x.com/android/forum/attachments/28-05-_2016_19-11-57-jpg.44477/

**Reply by IBBoard** [reply]
“A Scrim is a semi-transparent gradient layer that helps Text appear more readable against backgrounds.” 

Or just keep the header bar and don't worry about putting text over an image 😁

Like I said before, I think that headerbar-less version might work fine on a phone (where you often don't have a header bar anyway, because there's no close button etc). But on a desktop it's just out of place. I'm sure it's pretty a lot of the time, and cool that we can do it, but "consistent and usable" >>>> "pretty".

**Reply by CodedOre** [reply]
You have a point there too.

So, I'm going to implement an classic header bar to the display. However, I might keep the code for the inline header. 

I already thought about adding an hidden setting (only available through dconf) so you could move the `MediaDisplay` into the main window, as this would be better on phones and as I find it a massive pain to resize an modal window.
So, I might do this with the header as well, using a normal header bar as default with an hidden option for the likes of me and the phone distribution maintainers.

**Reply by jannuary** [reply]
> I now have (mostly) implemented the UI for the profile information:
> 
> ![current-profile-display](https://user-images.githubusercontent.com/47981497/145839336-98b8a0eb-6efe-4007-997a-26c19be80936.png)

Are we not doing AdwAvatar here? :/

**Reply by CodedOre** [reply]
Our [UserAvatar](https://github.com/CodedOre/NewCaw/blob/main/src/Widgets/UserAvatar.vala) uses `AdwAvatar` in the background, with an additional css class to make it the rounding optional.

See https://github.com/IBBoard/cawbird/issues/199
